### PR TITLE
Switching coverage to a better mechanism

### PR DESCRIPTION
**Pull request by ianthehat**
This works by tracking all coverage variables from the full transitive dependacy
set, which allows multiple packages to be covered at once.
This also normalises the names of cover variables so that two separate covers of
the same file will produce actions that merge.
It also tracks the cover variables directly, rather than inferring them from the
file names, which also means the test generator does not need to process the
generated files.

@linuxerwang you might want to check these changes, as it's based heavily on the work you did.

### CODE REVIEW

File: go/private/test.bzl [hunk 1]
```diff
@@ -41,21 +41,30 @@ def _go_test_impl(ctx):
...
+      '--rundir',
+      run_dir,
+      '--output',
+      main_go.path,
+  ]
+  cover_vars = []
+  covered_libs = []
+  for golib in depset([golib]) + golib.transitive:
+    if golib.cover_vars:
+        covered_libs += [golib]
```
Comment (jayconrod) [inline go/private/test.bzl]: Nit: 2-space indent

File: go/private/test.bzl [hunk 2]
```diff
@@ -66,7 +75,7 @@ def _go_test_impl(ctx):
       library = None,
       want_coverage = False,
       importpath = ctx.label.name + "~testmain~",
-      golibs = [golib],
+      golibs = [golib] + covered_libs,
```
Comment (jayconrod) [inline go/private/test.bzl]: We'll end up passing `golib` in multiple times here. This could go in another CL, but I think `emit_library_actions` should accept a depset instead of a list for this parameter.

File: go/private/test.bzl (hunk 1 above)
Comment (ianthehat) [inline go/private/test.bzl]: Done

File: go/private/test.bzl (hunk 2 above)
Comment (ianthehat) [inline go/private/test.bzl]: It actually does not matter because emit_library_actions immediately collects the libs into depsets internally (direct and transitive)

File: go/private/library.bzl [hunk 3]
```diff
@@ -272,8 +273,9 @@ def _emit_go_cover_action(ctx, out_object, sources):
       outputs += [src]
       continue
 
-    cover_var = "GoCover_%d" % count
-    out = ctx.new_file(out_object, out_object.basename + '_' + src.basename[:-3] + '_' + cover_var + '.cover.go')
+    cover_var = "Cover_" + src.basename[:-3].replace("-", "_").replace(".", "_")
+    cover_vars += ["{}={}".format(cover_var,src.short_path)]
```
Comment (linuxerwang) [inline go/private/library.bzl]: nit: one space after ",".

File: go/private/test.bzl [hunk 4]
```diff
@@ -41,21 +41,30 @@ def _go_test_impl(ctx):
...
   main_go = ctx.new_file(ctx.label.name + "_main_test.go")
+  arguments = [
+      '--package',
+      golib.importpath,
+      '--rundir',
+      run_dir,
+      '--output',
+      main_go.path,
+  ]
+  cover_vars = []
```
Comment (linuxerwang) [inline go/private/test.bzl]: Seem not used.

File: go/tools/builders/generate_test_main.go [hunk 5]
```diff
@@ -93,6 +62,7 @@ import (
 	"flag"
 	"log"
 	"os"
+	"fmt"
```
Comment (linuxerwang) [inline go/tools/builders/generate_test_main.go]: Move up to line 63.
